File: crates/opensymphony-gateway-schema/src/approval.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use super::version::SchemaVersion;

/// A live ACP callback. This is deliberately ephemeral: a restart cannot
/// restore an operator decision without the same live RPC responder.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct OperatorInteraction {
    pub request_id: String,
    pub run_id: String,
    pub issue_id: String,
    pub issue_identifier: String,
    pub session_id: String,
    pub generation: u64,
    /// Generated public binding token; the original peer RPC ID stays private
    /// inside the ACP responder. Clients echo this token unchanged.
    pub rpc_id: String,
    pub kind: OperatorInteractionKind,
    pub title: String,
    pub options: Vec<OperatorOption>,
    pub questions: Vec<OperatorQuestion>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plan: Option<String>,
    pub requested_at: DateTime<Utc>,
    pub expires_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum OperatorInteractionKind {
    Permission,
    Question,
    PlanApproval,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct OperatorOption {
    pub id: String,
    pub label: String,
    pub kind: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct OperatorQuestion {
    pub id: String,
    pub prompt: String,
    pub options: Vec<OperatorOption>,
    pub allow_multiple: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct OperatorQuestionAnswer {
    pub question_id: String,
    pub selected_option_ids: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum OperatorAnswer {
    Permission {
        option_id: String,
    },
    Question {
        answers: Vec<OperatorQuestionAnswer>,
    },
    Plan {
        accepted: bool,
    },
    Decline,
    Cancel,
}

/// Approval request exposed by the gateway for human-in-the-loop actions.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ApprovalRequest {
    pub schema_version: SchemaVersion,
    pub approval_id: String,
    pub run_id: String,
    pub issue_id: String,
    pub kind: ApprovalKind,
    pub title: String,
    pub description: String,
    pub proposed_action: Option<serde_json::Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub operator_interaction: Option<OperatorInteraction>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub actor: Option<ApprovalActor>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub target_context: Option<ApprovalTargetContext>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub risk_summary: Option<ApprovalRiskSummary>,
    pub requested_at: DateTime<Utc>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<DateTime<Utc>>,
    pub status: ApprovalStatus,
    pub correlation_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub decided_at: Option<DateTime<Utc>>,
}

/// Actor requesting an approval.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ApprovalActor {
    pub actor_id: String,
    pub actor_kind: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
}

/// Context the approval applies to (file, command, issue, or run).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ApprovalTargetContext {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub file_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub command: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub issue_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub issue_identifier: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub run_id: Option<String>,
}

/// Risk level for an approval request.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ApprovalRiskLevel {
    Low,
    Medium,
    High,
    Unknown,
}

/// Risk summary associated with an approval request.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ApprovalRiskSummary {
    pub level: ApprovalRiskLevel,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub reasons: Vec<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ApprovalKind {
    ToolUse,
    FileWrite,
    CommandExecution,
    PlanPublish,
    Custom,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ApprovalStatus {
    Pending,
    Approved,
    Rejected,
    Expired,
    Cancelled,
    Passed,
    Failed,
}

// NOTE: ActionReceipt and ActionStatus are defined in the action module.
// Use `opensymphony_gateway_schema::action::{ActionReceipt, ActionStatus}`.

File: packages/gateway-schema/src/approval.ts
import type { SchemaVersion } from "./version.js";

export type ApprovalKind =
  | "tool_use"
  | "file_write"
  | "command_execution"
  | "plan_publish"
  | "custom";

export type ApprovalStatus =
  | "pending"
  | "approved"
  | "rejected"
  | "expired"
  | "cancelled"
  | "passed"
  | "failed";

/** Actor that requested or approved an action. */
export interface ApprovalActor {
  actor_id: string;
  actor_kind: string;
  display_name?: string;
}

/** Target context for an approval request (file, command, issue, run). */
export interface ApprovalTargetContext {
  file_path?: string;
  command?: string;
  issue_id?: string;
  issue_identifier?: string;
  run_id?: string;
}

/** Risk level for an approval request. */
export type ApprovalRiskLevel = "low" | "medium" | "high" | "unknown";

/** Risk summary for an approval request. */
export interface ApprovalRiskSummary {
  level: ApprovalRiskLevel;
  reasons: string[];
}

/** Approval request for human-in-the-loop actions. */
export interface ApprovalRequest {
  schema_version: SchemaVersion;
  approval_id: string;
  run_id: string;
  issue_id: string;
  kind: ApprovalKind;
  title: string;
  description: string;
  proposed_action?: unknown;
  /** Actor who requested the approval. */
  actor?: ApprovalActor;
  /** Target context (file, command, issue, run) for the request. */
  target_context?: ApprovalTargetContext;
  /** Risk summary when available. */
  risk_summary?: ApprovalRiskSummary;
  requested_at: string;
  /** ISO timestamp when the approval was decided, if already decided. */
  decided_at?: string;
  expires_at?: string;
  status: ApprovalStatus;
  correlation_id: string;
}

/** Status of an action after gateway validation. */
export type ActionStatus = "accepted" | "rejected";

/** Expected follow-up event type after an action. */
export type ExpectedFollowup =
  | "state_transition"
  | "run_lifecycle"
  | "action_completion"
  | "journal_update"
  | "task_graph_update";

/** Permission check result for hosted mode. */
export interface PermissionResult {
  allowed: boolean;
  required_role: string;
  evaluated: boolean;
}

/** Action receipt returned after a mutation. */
export interface ActionReceipt {
  schema_version: SchemaVersion;
  action_id: string;
  correlation_id: string;
  status: ActionStatus;
  reason?: string;
  expected_followup: ExpectedFollowup[];
  result?: unknown;
  issued_at: string;
  /** Hosted-mode permission check result; omitted in local mode. */
  permission?: PermissionResult;
}

/** @deprecated Use `ActionStatus` instead. */
export type ActionReceiptStatus = ActionStatus;
